### PR TITLE
VRT vault

### PR DESCRIPTION
**Pull request by defistar**
## Description

- VRTVault contracts to let user deposit VRT for a fixed Interest Rate
- No LockTime for the funds
- User can deposit as many time as they wish to
- Interest is accrued based on blocks passed from the time since deposit was made
- Any existing deposit gets interest accrued and transferred to the User
- InterestRate is immutable and set once during contract deployment

## VRTVault Public and external functions
1. Deposit
2. Withdraw
3. Claim
4. getAccruedInterest
5. getBlockNumber

## VRTVault Admin functions

1. getAdmin
2. burnAdmin
3. setNewAdmin
4. setVenusInfo
5. _become
6. _setPendingImplementation
7. _acceptImplementation
8. _setPendingAdmin
9. _acceptAdmin

### CODE REVIEW

**Review by kkirka — CHANGES_REQUESTED** [review]

File: contracts/VRTVault/VRTVault.sol [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+            uint256 accruedInterest = computeAccruedInterest(user.totalPrincipalAmount, user.accrualStartBlockNumber);
+
+            user.totalPrincipalAmount = user.totalPrincipalAmount.add(depositAmount);
+
+            if(accruedInterest > 0){
+                uint256 vrtBalance = vrt.balanceOf(address(this));
+                require(vrtBalance >= accruedInterest, "Failed to transfer accruedInterest, Insufficient VRT in Vault.");
+                user.totalInterestAmount = user.totalInterestAmount.add(accruedInterest);
+                emit Claim(userAddress, accruedInterest);
+                vrt.safeTransferFrom(address(this), user.userAddress, accruedInterest);
```
[the remark this answers is not in the document]
Comment (defistar) [reply contracts/VRTVault/VRTVault.sol:87]: @kkirka will change this. to

                vrt.safeTransfer(user.userAddress, accruedInterest);

File: contracts/VRT/VRT.sol [hunk 2]
```diff
@@ -0,0 +1,301 @@
+pragma solidity ^0.5.16;
```
[the remark this answers is not in the document]
Comment (defistar) [reply contracts/VRT/VRT.sol]: Hi @kkirka 
i have refactored `VRT` to contracts/Tokens

shall i refactor XVS and SXP in this PR?
or
shall i make it as a new PR?

[the remark this answers is not in the document]
Comment (kkirka) [reply contracts/VRT/VRT.sol]: Definitely not in this PR, let's have a new one. Please create a new task for this and schedule it to the next sprint.

[the remark this answers is not in the document]
Comment (defistar) [reply contracts/VRT/VRT.sol]: @kkirka  I made changes only for VRT 
  - move VRT to directory Tokens
  - VRT will refer to Tokenlock and Owned contracts  in Utils directory
  
shall i retain them or revert?

** VRT contract never exist in contracts repo, i got its code from mainnet, and created the files as-is

File: contracts/VRTVault/VRTVault.sol [hunk 3]
```diff
@@ -49,9 +49,10 @@ contract VRTVault is WithAdmin, ReentrancyGuard, VRTVaultStorage {
      * @notice Deposit VRT to VRTVault for a fixed-interest-rate
      * @param depositAmount The amount to deposit to vault
      */
-    function deposit(address userAddress, uint256 depositAmount) public nonReentrant nonZeroAddress(userAddress) {
+    function deposit(uint256 depositAmount) public nonReentrant nonZeroAddress(msg.sender) {
```
Comment (kkirka) [inline contracts/VRTVault/VRTVault.sol:52]: No need to check msg.sender for zero

File: tests/VRTVault/vrtVaultTest.js [hunk 4]
```diff
@@ -219,21 +217,29 @@ describe('XVSVault', () => {
       const accrualStartBlockNumber = await getAccrualStartBlockNumber(vrtVault, user1);
       const expectedAccruedInterest = await calculateAccruedInterest(vrtDepositAmount, accrualStartBlockNumber, currentBlockNumber)
 
-      const vrtClaimTransaction = await send(vrtVault, "claim", [user1], { from: user1 });
+      //claim
+      const tokenBalance_Before_Claim = await getBep20balance(vrt, user1);
+      const tokenBalance_Of_Vault_Before_Claim = await getBep20balance(vrt, vrtVaultAddress);
```
Comment (kkirka) [inline tests/VRTVault/vrtVaultTest.js:222]: The rest of the project uses camelCase without underscores for variable names 🤔

**Review by kkirka — CHANGES_REQUESTED** [review]

File: contracts/VRTVault/VRTVault.sol [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+    }
+
+    /**
+     * @dev Set the current admin to new address
+     */
+    function setNewAdmin(address newAdmin) public onlyAdmin {
+        require(newAdmin != address(0), "new owner is the zero address");
+        emit AdminTransfered(admin, newAdmin);
+        admin = newAdmin;
+    }
```
Comment (kkirka) [inline contracts/VRTVault/VRTVault.sol:219]: We can use proxy functions to get and set admins, do we really need them in the proxy? `setNewAdmin` removes two-step admin change, and emits an event different from the one that proxy emits.

File: contracts/VRTVault/VRTVault.sol [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+    function _become(VRTVaultProxy vrtVaultProxy) public {
+        require(msg.sender == vrtVaultProxy.admin(), "only proxy admin can change brains");
+        vrtVaultProxy._acceptImplementation();
+    }
+
+    function setVenusInfo(address _vrt) public onlyAdmin {
+        vrt = IBEP20(_vrt);
+
+        _notEntered = true;
+    }
```
Comment (kkirka) [inline contracts/VRTVault/VRTVault.sol:232]: We set these in `initialize`, looks like we don't need this function

File: contracts/VRTVault/VRTVault.sol [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+
+    /// @notice Event emitted when Admin withdraw BEP20 token from contract
+    event WithdrawToken(address indexed tokenAddress, address indexed receiver, uint256 amount);
+
+    /// @notice Event emitted when accruedInterest is claimed
+    event Claim(address indexed user, uint256 interestAmount);
+
+    constructor() public {
+        admin = msg.sender;
+    }
```
Comment (kkirka) [inline contracts/VRTVault/VRTVault.sol:31]: Why do we need the constructor at all here? To simplify testing?

File: contracts/VRTVault/VRTVaultStorage.sol [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+    /**
+    * @notice Pending administrator for this contract
+    */
+    address public pendingAdmin;
+
+    /**
+    * @notice Active brains of VRT Vault
+    */
+    address public vrtVaultImplementation;
```
Comment (kkirka) [inline contracts/VRTVault/VRTVaultStorage.sol:19]: bscscan does not recognize `vrtVaultImplementation` as a pointer to the implementation contract. Let's use a conventional name here:

```suggestion
    address public implementation;
```

File: contracts/VRTVault/VRTVaultStorage.sol [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+    /**
+    * @notice Pending brains of VAI Vault
+    */
+    address public pendingVRTVaultImplementation;
```
Comment (kkirka) [inline contracts/VRTVault/VRTVaultStorage.sol:24]: And here as well:

```suggestion
    address public pendingImplementation;
```

File: contracts/VRTVault/VRTVault.sol (hunk 5 above)
Comment (defistar) [reply contracts/VRTVault/VRTVault.sol:219]: @kkirka yeah unused.  these were the functions added from XVSVault / VAIVault contracts
i will remove these.

File: contracts/VRTVault/VRTVault.sol (hunk 7 above)
Comment (defistar) [reply contracts/VRTVault/VRTVault.sol:31]: @kkirka for testing

File: contracts/VRTVault/VRTVault.sol (hunk 6 above)
Comment (defistar) [reply contracts/VRTVault/VRTVault.sol:232]: Removed the function

File: contracts/VRTVault/VRTVaultStorage.sol (hunk 8 above)
Comment (defistar) [reply contracts/VRTVault/VRTVaultStorage.sol:19]: renamed to implementation

File: contracts/VRTVault/VRTVaultStorage.sol (hunk 9 above)
Comment (defistar) [reply contracts/VRTVault/VRTVaultStorage.sol:24]: renamed to pendingImplementation

**Comment by kkirka** [comment]
Task linked: [CU-22h3eka VRT vault](https://app.clickup.com/t/22h3eka)

**Review by kkirka — APPROVED** [review]

**Comment by kkirka** [comment]
Please rebase --autosquash before merging, and wait for @RoseRompkxm review as well

**Review by RoseRompkxm — APPROVED** [review]

**Pull request merged**